queue/abstract/driver/fifo.lua

```diff
@@ -66,9 +66,36 @@ function method.normalize_task(self, task)
     return task
 end
 
+-- check if mvcc is enabled
+local function check_mvcc_state()
+    local cfg = box.cfg
+    if cfg['memtx_use_mvcc_engine'] == nil  or cfg['memtx_use_mvcc_engine'] == false then
+        return false
+    end
+
+    return true
+end
+
 -- put task in space
 function method.put(self, data, opts)
-    local max = self.space.index.task_id:max()
+    local max
+
+    -- taking the maximum of the index is an implicit transactions, so it is
+    -- always done with 'read-confirmed' mvcc isolation level.
+    -- It can lead to errors when trying to make parallel 'put' calls with mvcc enabled.
+    -- It is hapenning because 'max' for several puts in parallel will be the same since
+    -- read confirmed isolation level makes visible all transactions that finished the commit.
+    -- To fix it we wrap it with box.begin/commit and set right isolation level.
+    -- See https://github.com/tarantool/queue/issues/207
+    -- See https://www.tarantool.io/ru/doc/latest/concepts/atomic/txn_mode_mvcc/
+    if check_mvcc_state() then
+        box.begin({txn_isolation = 'read-committed'})
+        max = self.space.index.task_id:max()
+        box.commit()
```

Comment:
Can't we lean on some tarantool's facilities to autogenerate an identifier? For example:

* Use `sequence = true` index option and pass `nil` as an ID.
* Use uuid.

Reply:
Maybe just save counter in lua runtime.
Load from max() at init moment.

Reply:
Yes, that is other options to fix that issue that we take into account. Currently we disscused to stick to the box.begin/commit.

Reply:
> Can you, please, reveal the reasons why the other alternatives were declined?
> 
> The given solution has the problem highlighted by @olegrok with nested transactions at least.

Sure. Problem highlighted by @olegrok  doesn't bother us because of 2 points:
1. Usecases when nested transactions will be opened are rare. (+ it will only happen when mvcc is on)
2. We didn't give promises that in that api there will be no box.begin calls. (that change will not be braking)
So we don't see that it is an issue.

Regarding  @R-omk  suggestion, it will need a lot of additional re-writing of 'queue' logic such as switch in  replicasets for example.

Given that @olegrok issue does'nt bother us. Current fix is enough.

Reply:
> Maybe just save counter in lua runtime. Load from max() at init moment.

Yes, we considered that option, but in this case we need to update that counter, when the master switched (replicaset mode on) and the current fix looks more localized.

Reply:
> Can't we lean on some tarantool's facilities to autogenerate an identifier? For example:
> 
> * Use `sequence = true` index option and pass `nil` as an ID.
> * Use uuid.

Switching to `uuid` will require alter several underlying spaces.
About using sequences, I need some time to think about whether that would work and be a better alternative.